src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,1001 @@
+                sin_theta_o,
+                std::abs(std::cos(theta_o_R)),
+                std::sin(theta_o_R),
+                std::abs(std::cos(theta_o_TT)),
+                std::sin(theta_o_TT),
+                std::abs(std::cos(theta_o_TRT)),
+                std::sin(theta_o_TRT),
+                values->m_precomputed.m_v,
+                dphi,
+                values->m_precomputed.m_s,
+                gamma_o,
+                gamma_t,
+                ap_pdf,
+                bsdf_pdf);
+            
+            const float angle_bsdf = std::abs(wi.y);
+
+            if (angle_bsdf > 0)
+                bsdf_f /= angle_bsdf;
+
+            value.m_glossy = bsdf_f;
+            value.m_beauty = bsdf_f;
+
+            return bsdf_pdf;
+
+        }
+
+        float evaluate_pdf(
+            const void*                 data,
+            const bool                  adjoint,
+            const LocalGeometry&        local_geometry,
+            const Vector3f&             outgoing,
+            const Vector3f&             incoming,
+            const int                   modes) const override
+        {
+            if (!ScatteringMode::has_glossy(modes))
+                return 0.0f;
+
+            // Compute the BRDF value.
+            const InputValues* values = static_cast<const InputValues*>(data);
+
+            const Vector3f wo = local_geometry.m_shading_basis.transform_to_local(outgoing);
+            const Vector3f wi = local_geometry.m_shading_basis.transform_to_local(incoming);
+
+            // Compute outgoing terms.
+            const float sin_theta_o = wo.x;
+            const float cos_theta_o = std::sqrt(std::max(0.0f, 1.0f - sin_theta_o * sin_theta_o));
+            const float phi_o = std::atan2(wo.y, wo.z);
+            const float gamma_o = std::asin(clamp(values->m_precomputed.m_h, -1.0f, 1.0f));
+
+            // Compute incoming terms.
+            const float sin_theta_i = wi.x;
+            const float cos_theta_i = std::sqrt(std::max(0.0f, 1.0f - sin_theta_i * sin_theta_i));
+            const float phi_i = std::atan2(wi.y, wi.z);
+
+            // Compute terms for the refracted ray.
+            const float sin_theta_t = sin_theta_o / values->m_eta;
+            const float cos_theta_t = std::sqrt(std::max(0.0f, 1.0f - sin_theta_t * sin_theta_t));
+
+            // Compute the modified refraction coefficient.
+            const float eta_p = std::sqrt(values->m_eta * values->m_eta - sin_theta_o * sin_theta_o) / cos_theta_o;
+            const float sin_gamma_t = values->m_precomputed.m_h / eta_p;
+            const float cos_gamma_t = std::sqrt(std::max(0.0f, 1.0f - sin_gamma_t * sin_gamma_t));
+            const float gamma_t = std::asin(clamp(sin_gamma_t, -1.0f, 1.0f));
+            const float theta_o = std::asin(clamp(sin_theta_o, -1.0f, 1.0f));
+
+            // Deviation of outgoing angle due to cuticular scales.
+            const float ang_in_radians = deg_to_rad<float>(values->m_alpha);
```

Comment:
Precompute in `on_frame_begin()`.

Reply:
I added some values to the precomputed input values that reduce the amount of recalculation.